drivers/spi/spi_gecko.c

```diff
@@ -0,0 +1,330 @@
+#endif
+
+	/* Enable USART clock */
+	CMU_ClockEnable(config->clock, true);
+
+	/* Init USART */
+	USART_InitSync(config->base, &usartInit);
+
+	/* Initialize USART pins */
+	spi_gecko_init_pins(dev);
+
+	/* Enable the peripheral */
+	config->base->CMD = (uint32_t) usartEnable;
+
+	return 0;
+}
+
+static int spi_gecko_transceive(struct device *dev,
+			  const struct spi_config *config,
+			  const struct spi_buf_set *tx_bufs,
+			  const struct spi_buf_set *rx_bufs)
+{
+	u16_t control = 0;
+
+	spi_config(dev, config, &control);
+	spi_context_buffers_setup(&DEV_DATA(dev)->ctx, tx_bufs, rx_bufs, 1);
+	spi_gecko_xfer(dev, config);
+	return 0;
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int spi_gecko_transceive_async(struct device *dev,
+			  const struct spi_config *config,
+			  const struct spi_buf_set *tx_bufs,
+			  const struct spi_buf_set *rx_bufs,
+			  struct k_poll_signal *async)
+{
+	return -ENOTSUP;
+}
+#endif /* CONFIG_SPI_ASYNC */
+
+static int spi_gecko_release(struct device *dev,
+			     const struct spi_config *config)
+{
+	const struct spi_gecko_config *gecko_config = dev->config->config_info;
+
+	if (!(gecko_config->base->STATUS & USART_STATUS_TXIDLE)) {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+/* Device Instantiation */
+static struct spi_driver_api spi_gecko_api = {
+	.transceive = spi_gecko_transceive,
+#ifdef CONFIG_SPI_ASYNC
+	.transceive_async = spi_gecko_transceive_async,
+#endif /* CONFIG_SPI_ASYNC */
+	.release = spi_gecko_release,
+};
+
+#define SPI_INIT2(n, usart)				    \
+	static struct spi_gecko_data spi_gecko_data_##n = { \
+		SPI_CONTEXT_INIT_LOCK(spi_gecko_data_##n, ctx), \
+		SPI_CONTEXT_INIT_SYNC(spi_gecko_data_##n, ctx), \
+	}; \
+	static struct spi_gecko_config spi_gecko_cfg_##n = { \
+	    .base = (USART_TypeDef *) \
+		 DT_INST_##n##_SILABS_GECKO_USART_SPI_BASE_ADDRESS, \
+	    .clock = CLOCK_USART(usart), \
+	    .pin_rx = { DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_RX_1, \
+			DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_RX_2, \
+			gpioModeInput, 1},				\
+	    .pin_tx = { DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_TX_1, \
+			DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_TX_2, \
+			gpioModePushPull, 1},				\
+	    .pin_clk = { DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_CLK_1, \
+			DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_CLK_2, \
+			gpioModePushPull, 1},				\
+	    .loc_rx = DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_RX_0, \
+	    .loc_tx = DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_TX_0, \
+	    .loc_clk = DT_INST_##n##_SILABS_GECKO_USART_SPI_LOCATION_CLK_0, \
+	}; \
+	DEVICE_AND_API_INIT(spi_##n, \
+			DT_INST_##n##_SILABS_GECKO_USART_SPI_LABEL, \
+			spi_gecko_init, \
+			&spi_gecko_data_##n, \
+			&spi_gecko_cfg_##n, \
+			POST_KERNEL, \
+			CONFIG_SPI_INIT_PRIORITY, \
+			&spi_gecko_api)
+
+#define SPI_ID(n) DT_INST_##n##_SILABS_GECKO_USART_SPI_PERIPHERAL_ID
+
+#define SPI_INIT(n) SPI_INIT2(n, SPI_ID(n))
+
+#ifdef DT_INST_0_SILABS_GECKO_USART_SPI_LABEL
+
+SPI_INIT(0);
```

Comment:
Since we have a macro already we could add a few more instances. efr32mg12 has 4 USART modules.